### PR TITLE
Improve the Event Record

### LINKED ISSUE #157: Messages that are sent because of the user going into an End Portal are spammed
When going to an unlink portal the plugin would put 

> You do not have access to go here

 in the chat. Going into an endportal however spams the message like 50 times a second. 

Pls add an option to diasble the message tq.

### LINKED ISSUE #158: Error when bounceback is disabled and no world is linked
If bounceback is disabled and no world is linked, when a user enters a portal they aren't bounced  back. So the PlayerPortalEvent is still fired, and since no world is linked we get this error:
```[15:53:22] [Server thread/ERROR]: Could not pass event PlayerPortalEvent to Multiverse-NetherPortals v4.1.1-SNAPSHOT-bUNKNOWN
java.lang.IllegalArgumentException: Cannot use null location!
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:122) ~[patched_1.15.2.jar:git-Paper-300]
	at org.bukkit.event.player.PlayerMoveEvent.validateLocation(PlayerMoveEvent.java:97) ~[patched_1.15.2.jar:git-Paper-300]
	at org.bukkit.event.player.PlayerMoveEvent.setTo(PlayerMoveEvent.java:92) ~[patched_1.15.2.jar:git-Paper-300]
	at com.onarandombox.MultiverseNetherPortals.listeners.MVNPPlayerListener.onPlayerPortal(MVNPPlayerListener.java:95) ~[?:?]
	at com.destroystokyo.paper.event.executor.asm.generated.GeneratedEventExecutor72.execute(Unknown Source) ~[?:?]
	at org.bukkit.plugin.EventExecutor.lambda$create$1(EventExecutor.java:69) ~[patched_1.15.2.jar:git-Paper-300]
	at co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:80) ~[patched_1.15.2.jar:git-Paper-300]
	at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70) ~[patched_1.15.2.jar:git-Paper-300]
	at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:607) ~[patched_1.15.2.jar:git-Paper-300]
	at net.minecraft.server.v1_15_R1.EntityPlayer.a(EntityPlayer.java:857) ~[patched_1.15.2.jar:git-Paper-300]
	at net.minecraft.server.v1_15_R1.BlockEnderPortal.a(BlockEnderPortal.java:34) ~[patched_1.15.2.jar:git-Paper-300]
	at net.minecraft.server.v1_15_R1.IBlockData.a(IBlockData.java:224) ~[patched_1.15.2.jar:git-Paper-300]
	at net.minecraft.server.v1_15_R1.Entity.checkBlockCollisions(Entity.java:1010) ~[patched_1.15.2.jar:git-Paper-300]
	at net.minecraft.server.v1_15_R1.Entity.move(Entity.java:725) ~[patched_1.15.2.jar:git-Paper-300]
	at net.minecraft.server.v1_15_R1.EntityLiving.e(EntityLiving.java:2270) ~[patched_1.15.2.jar:git-Paper-300]
	at net.minecraft.server.v1_15_R1.EntityHuman.e(EntityHuman.java:1559) ~[patched_1.15.2.jar:git-Paper-300]
	at net.minecraft.server.v1_15_R1.EntityLiving.movementTick(EntityLiving.java:2637) ~[patched_1.15.2.jar:git-Paper-300]
	at net.minecraft.server.v1_15_R1.EntityHuman.movementTick(EntityHuman.java:428) ~[patched_1.15.2.jar:git-Paper-300]
	at net.minecraft.server.v1_15_R1.EntityLiving.tick(EntityLiving.java:2396) ~[patched_1.15.2.jar:git-Paper-300]
	at net.minecraft.server.v1_15_R1.EntityHuman.tick(EntityHuman.java:165) ~[patched_1.15.2.jar:git-Paper-300]
	at net.minecraft.server.v1_15_R1.EntityPlayer.playerTick(EntityPlayer.java:442) ~[patched_1.15.2.jar:git-Paper-300]
	at net.minecraft.server.v1_15_R1.PlayerConnection.tick(PlayerConnection.java:159) ~[patched_1.15.2.jar:git-Paper-300]
	at net.minecraft.server.v1_15_R1.NetworkManager.a(NetworkManager.java:343) ~[patched_1.15.2.jar:git-Paper-300]
	at net.minecraft.server.v1_15_R1.ServerConnection.c(ServerConnection.java:140) ~[patched_1.15.2.jar:git-Paper-300]
	at net.minecraft.server.v1_15_R1.MinecraftServer.b(MinecraftServer.java:1344) ~[patched_1.15.2.jar:git-Paper-300]
	at net.minecraft.server.v1_15_R1.DedicatedServer.b(DedicatedServer.java:430) ~[patched_1.15.2.jar:git-Paper-300]
	at net.minecraft.server.v1_15_R1.MinecraftServer.a(MinecraftServer.java:1173) ~[patched_1.15.2.jar:git-Paper-300]
	at net.minecraft.server.v1_15_R1.MinecraftServer.run(MinecraftServer.java:962) ~[patched_1.15.2.jar:git-Paper-300]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_241]
```

### PR DESCRIPTION
**Pull request by nicegamer7**
This PR fixes #157 by reworking how the event record works. Additionally, there is an added config option to disable the messages sent when there is no destination world.

This PR also fixes #158.

### CODE REVIEW

**Comment by nicegamer7** [comment]
It seems that #158 is a similar issue to #110, so this PR would close it too.

**Comment by nicegamer7** [comment]
I should add that this PR implements the same fix as #117, and thus supersedes it. This is ready for review!

**Comment by nicegamer7** [comment]
The commit I just pushed fixes #164. I'm surprised no one else caught this (\*face palm\*). Below is a test build of this PR.

[Multiverse-NetherPortals-4.1.1-SNAPSHOT.zip](https://github.com/Multiverse/Multiverse-NetherPortals/files/4748584/Multiverse-NetherPortals-4.1.1-SNAPSHOT.zip)

**Comment by nicegamer7** [comment]
I decided to revisit my fix for bounceback direction from #156 and I'm glad I did! I noticed that my fix from #156 only fixed bounceback direction when entering a portal from the South and from the West, but North and East remained broken! I amended that by adding 0.5 to the location of the portal block, thus checking the player's location relative to the center of the block.

With that, this PR is now ready for review!

**Comment by nicegamer7** [comment]
I spoke too soon... I knew I was forgetting something.

It turns out, since the runnable is not async, there is no need to use Concurrent Maps/Sets. *Now* this PR is ready for review.

**Comment by nicegamer7** [comment]
One final commit to fix #165. A build of this PR is attached.

[Multiverse-NetherPortals-4.1.1-SNAPSHOT.zip](https://github.com/Multiverse/Multiverse-NetherPortals/files/4768584/Multiverse-NetherPortals-4.1.1-SNAPSHOT.zip)

**Comment by Dnaleor77** [comment]
> One final commit to fix #165. A build of this PR is attached.
> 
> [Multiverse-NetherPortals-4.1.1-SNAPSHOT.zip](https://github.com/Multiverse/Multiverse-NetherPortals/files/4768584/Multiverse-NetherPortals-4.1.1-SNAPSHOT.zip)

I will give it a try. Without MV-Nether Portals, all the end portals point towards the same world_the_end

**Comment by thenullmatrix** [comment]
Yep I get the same crash log when someone uses the return portal. also the end gateways do not work as well. This only occurs in the other end worlds not in the main server end world.

**Comment by nicegamer7** [comment]
Just for clarity, the two above comments were related to https://github.com/Multiverse/Multiverse-Core/issues/2332 (not regressions).

File: src/main/java/com/onarandombox/MultiverseNetherPortals/commands/LinkCommand.java [hunk 1]
```diff
@@ -41,7 +41,11 @@ public void runCommand(CommandSender sender, List<String> args) {
         String toWorldString;
         PortalType type;
         Player p;
-        type = PortalType.parse(args.get(0).toUpperCase());
+
+        if (args.get(0).equalsIgnoreCase("END")) type = PortalType.ENDER;
+        else if (args.get(0).equalsIgnoreCase("NETHER")) type = PortalType.NETHER;
+        else type = null;
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseNetherPortals/commands/LinkCommand.java:47]: Please use braces here.

File: src/main/java/com/onarandombox/MultiverseNetherPortals/commands/UnlinkCommand.java [hunk 2]
```diff
@@ -34,13 +34,17 @@ public void runCommand(CommandSender sender, List<String> args) {
...
+        String fromWorldString;
+        String toWorldString;
         PortalType type;
-        Player p = null;
-        type = PortalType.parse(args.get(0));
+        Player p;
+
+        if (args.get(0).equalsIgnoreCase("END")) type = PortalType.ENDER;
+        else if (args.get(0).equalsIgnoreCase("NETHER")) type = PortalType.NETHER;
+        else type = null;
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseNetherPortals/commands/UnlinkCommand.java:46]: Braces, please.

File: src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java [hunk 3]
```diff
@@ -49,41 +54,58 @@ public MVNPEntityListener(MultiverseNetherPortals plugin) {
         this.worldManager = this.plugin.getCore().getMVWorldManager();
         this.pt = new PermissionTools(this.plugin.getCore());
         this.playerErrors = new HashMap<String, Date>();
-        this.eventRecord = new HashMap<String, Location>();
         this.messaging = this.plugin.getCore().getMessaging();
         this.locationManipulation = this.plugin.getCore().getLocationManipulation();
-
+        this.eventRecord = new HashMap<>();
+        this.eventRecord.put(PortalType.ENDER, new HashSet<>());
+        this.eventRecord.put(PortalType.NETHER, new HashSet<>());
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java:61]: This is fine. However, to me this smells like a good place for a purpose built class.

File: src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java [hunk 4]
```diff
@@ -49,41 +54,58 @@ public MVNPEntityListener(MultiverseNetherPortals plugin) {
...
     }
 
-    protected void shootPlayer(Player p, Block block, PortalType type) {
+    protected boolean shootPlayer(Player p, Block block, PortalType type) {
         if (!plugin.isUsingBounceBack()) {
-            this.plugin.log(Level.FINEST, "You said not to use bounce back so the player is free to walk into portal!");
-            return;
+            this.plugin.log(Level.FINEST, "Bounceback is disabled, so the player is free to walk into portals!");
+            return false;
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java:67]: A good change. I think even better might be `"Bounceback is disabled. The player is free to walk into the portal!"`

File: src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java [hunk 5]
```diff
@@ -49,41 +54,58 @@ public MVNPEntityListener(MultiverseNetherPortals plugin) {
...
         this.messaging = this.plugin.getCore().getMessaging();
         this.locationManipulation = this.plugin.getCore().getLocationManipulation();
-
+        this.eventRecord = new HashMap<>();
+        this.eventRecord.put(PortalType.ENDER, new HashSet<>());
+        this.eventRecord.put(PortalType.NETHER, new HashSet<>());
     }
 
-    protected void shootPlayer(Player p, Block block, PortalType type) {
+    protected boolean shootPlayer(Player p, Block block, PortalType type) {
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java:64]: Would you mind adding a javadoc here to help understand what the return value means?

File: src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java [hunk 6]
```diff
@@ -49,41 +54,58 @@ public MVNPEntityListener(MultiverseNetherPortals plugin) {
...
+            // Determine portal axis:
+            if (block.getRelative(BlockFace.EAST).getType() == Material.NETHER_PORTAL || block.getRelative(BlockFace.WEST).getType() == Material.NETHER_PORTAL) {
+                // we add 0.5 to the location of the block to get the center
+                if (p.getLocation().getZ() < block.getLocation().getZ() + 0.5) {
+                    debugMessage.append(" entered Nether Portal from the North");
+                    newVecZ = -1 * strength;
+                } else {
+                    debugMessage.append(" entered Nether Portal from the South");
+                    newVecZ = 1 * strength;
+                }
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java:89]: This method could do with some serious refactoring. It's very hard to follow what is going on here. (Not your fault)

File: src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java [hunk 7]
```diff
@@ -49,41 +54,58 @@ public MVNPEntityListener(MultiverseNetherPortals plugin) {
...
+            debugMessage.append(". They will be bounced back!");
+            p.teleport(p.getLocation().clone().add(newVecX, .2, newVecZ));
+            p.setVelocity(new Vector(newVecX, .6, newVecZ));
+            playerBounced = true;
         }
-        p.teleport(p.getLocation().clone().add(newVecX, .2, newVecZ));
-        p.setVelocity(new Vector(newVecX, .6, newVecZ));
+
+        this.plugin.log(Level.FINER, debugMessage.toString());
+        return playerBounced;
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java:108]: I'm just gonna trust this all correct cause I CBA to read and understand it.

File: src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java [hunk 8]
```diff
@@ -7,40 +7,45 @@
...
+    private final MultiverseNetherPortals plugin;
+    private final MVNameChecker nameChecker;
+    private final MVLinkChecker linkChecker;
+    private final MVWorldManager worldManager;
+    private final PermissionTools pt;
+    private final int cooldown = 250;
+    private final MultiverseMessaging messaging;
+    private final Map<String, Date> playerErrors;
+    private final LocationManipulation locationManipulation;
+    private final Map<PortalType, Set<Player>> eventRecord;
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java:45]: We definitely do not want to use `Player` objects here. Storing player objects in a persistent variable can cause a lot of issues (e.g. ghost players). It would probably be better to use UUID here.

File: src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java [hunk 9]
```diff
@@ -154,42 +176,49 @@ public void onEntityPortalEnter(EntityPortalEnterEvent event) {
...
         MultiverseWorld fromWorld = this.worldManager.getMVWorld(p.getLocation().getWorld().getName());
         MultiverseWorld toWorld = this.worldManager.getMVWorld(toLocation.getWorld().getName());
+
         if (fromWorld.getCBWorld().equals(toWorld.getCBWorld())) {
             // The player is Portaling to the same world.
             this.plugin.log(Level.FINER, "Player '" + p.getName() + "' is portaling to the same world.");
             return;
         }
         if (!pt.playerHasMoneyToEnter(fromWorld, toWorld, p, p, false)) {
-            System.out.println("BOOM");
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java:187]: WTF

File: src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java [hunk 10]
```diff
@@ -200,4 +229,12 @@ public void onEntityPortalEnter(EntityPortalEnterEvent event) {
...
-}
+
+    public void onEntityPortalExit(EntityPortalExitEvent event) {
+        if (event.getEntity() instanceof Player) {
+            Player p = (Player) event.getEntity();
+            this.eventRecord.get(PortalType.ENDER).remove(p);
+            this.eventRecord.get(PortalType.NETHER).remove(p);
+        }
+    }
+}
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java:240]: Is this suppose to be annotated with `@EventHandler`? Also, let's keep new lines at the end of files as a general rule.

File: src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPPlayerListener.java [hunk 11]
```diff
@@ -35,25 +36,19 @@ public MVNPPlayerListener(MultiverseNetherPortals plugin) {
         this.linkChecker = new MVLinkChecker(this.plugin);
     }
 
-    @EventHandler(priority = EventPriority.NORMAL)
+    @EventHandler(priority = EventPriority.NORMAL, ignoreCancelled = true)
     public void onPlayerPortal(PlayerPortalEvent event) {
-        if (event.isCancelled()) {
-            this.plugin.log(Level.FINEST, "PlayerPortalEvent was cancelled! NOT teleporting!");
-            return;
-        }
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPPlayerListener.java:43]: My only note about this is that we lose the logging. However, I'm not sure if this particular log line is very helpful.

File: src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPPlayerListener.java [hunk 12]
```diff
@@ -35,25 +36,19 @@ public MVNPPlayerListener(MultiverseNetherPortals plugin) {
...
         }
-        String currentWorld = currentLocation.getWorld().getName();
 
-        PortalType type = PortalType.END;
-        if (event.getFrom().getBlock().getType() == Material.NETHER_PORTAL) {
-            type = PortalType.NETHER;
+        PortalType type;
+        if (event.getCause() == PlayerTeleportEvent.TeleportCause.END_PORTAL) type = PortalType.ENDER;
+        else if (event.getCause() == PlayerTeleportEvent.TeleportCause.NETHER_PORTAL) type = PortalType.NETHER;
+        else return;
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPPlayerListener.java:49]: braces

File: src/main/java/com/onarandombox/MultiverseNetherPortals/runnables/PlayerTouchingPortalTask.java [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+        this.portalType = portalType;
+        this.player = player;
+    }
+
+    @Override
+    public void run() {
+        PortalType type = null;
+
+        if (player.getLocation().getBlock().getType() == Material.END_PORTAL) type = PortalType.ENDER;
+        else if (player.getLocation().getBlock().getType() == Material.NETHER_PORTAL) type = PortalType.NETHER;
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseNetherPortals/runnables/PlayerTouchingPortalTask.java:27]: braces

File: src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPPlayerListener.java [hunk 14]
```diff
@@ -62,44 +57,44 @@ public void onPlayerPortal(PlayerPortalEvent event) {
...
-            this.linkChecker.getNewTeleportLocation(event, currentLocation, linkedWorld);
+        if (currentWorld.equalsIgnoreCase(linkedWorld)) {
+            newTo = null;
+        } else if (linkedWorld != null) {
+            newTo = this.linkChecker.findNewTeleportLocation(currentLocation, linkedWorld, event.getPlayer());
         } else if (this.nameChecker.isValidNetherName(currentWorld)) {
             if (type == PortalType.NETHER) {
-                this.plugin.log(Level.FINER, "");
-                this.linkChecker.getNewTeleportLocation(event, currentLocation, this.nameChecker.getNormalName(currentWorld, PortalType.NETHER));
+                newTo = this.linkChecker.findNewTeleportLocation(currentLocation, this.nameChecker.getNormalName(currentWorld, PortalType.NETHER), event.getPlayer());
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPPlayerListener.java:69]: What's the diff between these two calls?

File: src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPPlayerListener.java [hunk 15]
```diff
@@ -62,44 +57,44 @@ public void onPlayerPortal(PlayerPortalEvent event) {
...
-        if (event.getTo() == null || event.getFrom() == null) {
-            return;
-        }
-        if (event.getFrom().getWorld().equals(event.getTo().getWorld())) {
-            // The player is Portaling to the same world.
-            this.plugin.log(Level.FINER, "Player '" + event.getPlayer().getName() + "' is portaling to the same world.  Ignoring.");
-            event.setTo(originalTo);
+
+        if (newTo != null) event.setTo(newTo);
+        else {
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPPlayerListener.java:88]: braces

File: src/main/java/com/onarandombox/MultiverseNetherPortals/runnables/PlayerTouchingPortalTask.java [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+import org.bukkit.Material;
+import org.bukkit.PortalType;
+import org.bukkit.entity.Player;
+import org.bukkit.scheduler.BukkitRunnable;
+
+import java.util.Map;
+import java.util.Set;
+
+public class PlayerTouchingPortalTask extends BukkitRunnable {
+    private final Map<PortalType, Set<Player>> eventRecord;
```
Comment (dumptruckman) [inline src/main/java/com/onarandombox/MultiverseNetherPortals/runnables/PlayerTouchingPortalTask.java:12]: As mentioned above, avoid using `Player` objects for this.

**Review by dumptruckman — CHANGES_REQUESTED** [review]
Thanks @nicegamer7 for the hard work on this. There's some issues that need to be resolved before it is ready but definitely a step in the right direction.

File: src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java (hunk 8 above)
Comment (nicegamer7) [reply src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java:45]: Good point.

File: src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java (hunk 9 above)
Comment (nicegamer7) [reply src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java:187]: Not a clue lol.

File: src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java (hunk 4 above)
Comment (nicegamer7) [reply src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java:67]: Yeah, I agree.

File: src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPPlayerListener.java (hunk 14 above)
Comment (nicegamer7) [reply src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPPlayerListener.java:78]: The difference is that `getNewTeleportLocation` uses `event.setTo()` to set the new location, but this caused issues where the new location could be set to `null` and cause an error (see #158). `findNewTeleportLocation` does the exact same thing except that it returns the value instead of setting it directly (then we can check if it's `null` before using it).

File: src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java (hunk 5 above)
Comment (nicegamer7) [reply src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java:65]: Sure, and I'll try to refactor it while I'm at it.

File: src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java (hunk 3 above)
Comment (nicegamer7) [reply src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java:62]: I think that'd be fine. We could also make the event record static then, which I think would be good (let me know what you think).

File: src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java (hunk 10 above)
Comment (nicegamer7) [reply src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java]: Good catch!

File: src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java (hunk 3 above)
Comment (dumptruckman) [reply src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java:62]: Nah. I like to avoid static for the most part with Bukkit stuff. If you use /reload it can lead to memory leaks.

Comment (nicegamer7) [reply src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java:62]: I see, good thing I asked.

Comment (nicegamer7) [reply src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java]: Just pushed it now 😄.

**Comment by HexedHero** [comment]
Can we please pull all Ricegamers changes, guy keeping this shit alive.

**Comment by nicegamer7** [comment]
That's kind of you, but patience is a virtue 😃.

File: src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java (hunk 5 above)
Comment (nicegamer7) [reply src/main/java/com/onarandombox/MultiverseNetherPortals/listeners/MVNPEntityListener.java]: I couldn't really find a good way to refactor it :/

But it's got a Javadoc now. Also, I lowered the strength because it was a bit too strong imo.

**Comment by benwoo1110** [comment]
🍚 Ricegamers ftw! ;)
